src/wmtk/io/EdgeMeshReader.cpp

```diff
@@ -0,0 +1,164 @@
+    : m_filename(filename)
+    , m_type(type)
+{}
+void EdgeMeshReader::read(
+    Eigen::Matrix<long, -1, -1>& edges,
+    Eigen::MatrixXd& vertices,
+    Eigen::MatrixXd& vertices_w,
+    Eigen::MatrixXd& vertices_texture,
+    Eigen::MatrixXd& vertices_normal,
+    Eigen::MatrixXd& vertices_parameter)
+{
+    std::vector<std::pair<long, long>> vec_edges;
+    std::vector<std::vector<double>> vec_vertices;
+    std::vector<double> vec_vertices_w;
+    std::vector<std::vector<double>> vec_vertices_texture;
+    std::vector<std::vector<double>> vec_vertices_normal;
+    std::vector<std::vector<double>> vec_vertices_parameter;
+    switch (m_type) {
+    case OBJ:
+        read_obj(
+            vec_edges,
+            vec_vertices,
+            vec_vertices_w,
+            vec_vertices_texture,
+            vec_vertices_normal,
+            vec_vertices_parameter);
+        break;
+    case OFF:;
+    default: throw std::runtime_error("Unknown data type!"); break;
+    }
+
+    edges.resize(vec_edges.size(), 2);
+    vertices.resize(vec_vertices.size(), 3);
+    vertices_w.resize(vec_vertices.size(), 1);
+    vertices_texture.resize(vec_vertices_texture.size(), 3);
+    vertices_normal.resize(vec_vertices_normal.size(), 3);
+    vertices_parameter.resize(vec_vertices_parameter.size(), 3);
+
+    for (size_t i = 0; i < edges.rows(); i++) {
+        edges(i, 0) = vec_edges[i].first;
+        edges(i, 1) = vec_edges[i].second;
+    }
+    for (size_t i = 0; i < vertices.rows(); i++) {
+        vertices(i, 0) = vec_vertices[i][0];
+        vertices(i, 1) = vec_vertices[i][1];
+        vertices(i, 2) = vec_vertices[i][2];
+        vertices_w(i, 0) = vec_vertices[i][3];
+    }
+    for (size_t i = 0; i < vertices_texture.rows(); i++) {
+        vertices_texture(i, 0) = vec_vertices_texture[i][0];
+        vertices_texture(i, 1) = vec_vertices_texture[i][1];
+        vertices_texture(i, 2) = vec_vertices_texture[i][2];
+    }
+    for (size_t i = 0; i < vertices_normal.rows(); i++) {
+        vertices_normal(i, 0) = vec_vertices_normal[i][0];
+        vertices_normal(i, 1) = vec_vertices_normal[i][1];
+        vertices_normal(i, 2) = vec_vertices_normal[i][2];
+    }
+    for (size_t i = 0; i < vertices_parameter.rows(); i++) {
+        vertices_parameter(i, 0) = vec_vertices_parameter[i][0];
+        vertices_parameter(i, 1) = vec_vertices_parameter[i][1];
+        vertices_parameter(i, 2) = vec_vertices_parameter[i][2];
+    }
+}
+
+void EdgeMeshReader::read_obj(
+    std::vector<std::pair<long, long>>& edges,
+    std::vector<std::vector<double>>& vertices,
+    std::vector<double>& vertices_w,
+    std::vector<std::vector<double>>& vertices_texture,
+    std::vector<std::vector<double>>& vertices_normal,
+    std::vector<std::vector<double>>& vertices_parameter)
+{
+    std::ifstream f(m_filename);
+    if (!f.is_open()) {
+        throw std::runtime_error("can't open file!");
+    }
+
+    std::string buffer;
+    while (!f.eof()) {

```

Comment:
Follow the concept from that stackoverflow answer:
https://stackoverflow.com/questions/7868936/read-file-line-by-line-using-ifstream-in-c

```
#include <sstream>
#include <string>

std::string line;
while (std::getline(infile, line))
{
    std::istringstream iss(line);
    int a, b;
    if (!(iss >> a >> b)) { break; } // error

    // process pair (a,b)
}
```

There is no need for using `f.eof()`.